services/versioncache/versioncache.go

```diff
@@ -0,0 +1,284 @@
+		serviceType := models.MySQLServiceType
+		services, err := models.FindServices(tx.Querier, models.ServiceFilters{ServiceType: &serviceType})
+		if err != nil {
+			return err
+		}
+
+		serviceIDs := make(map[string]struct{}, len(services))
+		for _, s := range services {
+			serviceIDs[s.ServiceID] = struct{}{}
+		}
+
+		serviceVersions, err := models.FindServicesSoftwareVersions(tx.Querier, nil)
+		if err != nil {
+			return err
+		}
+
+		// remove services software versions from the cache which are no longer exist
+		for _, sv := range serviceVersions {
+			if _, ok := serviceIDs[sv.ServiceID]; !ok {
+				if err := models.DeleteServiceSoftwareVersions(tx.Querier, sv.ID); err != nil {
+					return err
+				}
+			}
+		}
+
+		// add new services software versions to the cache
+		cacheServiceIDs := make(map[string]struct{}, len(serviceVersions))
+		for _, sv := range serviceVersions {
+			cacheServiceIDs[sv.ServiceID] = struct{}{}
+		}
+		for _, s := range services {
+			if _, ok := cacheServiceIDs[s.ServiceID]; !ok {
+				if _, err := models.CreateServiceSoftwareVersions(tx.Querier, models.CreateServiceSoftwareVersionsParams{
+					ServiceID: s.ServiceID,
+					Versions:  []models.SoftwareVersion{},
+					CheckAt:   time.Now().UTC(),
+				}); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+type prepareResults struct {
+	ID          string
+	CheckAfter  time.Duration
+	NeedsUpdate bool
+	AgentID     string
+}
+
+func (s *Service) prepareUpdateVersions() (*prepareResults, error) {
+	results := &prepareResults{CheckAfter: relaxDuration}
+
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		limit := 1
+		servicesVersions, err := models.FindServicesSoftwareVersions(tx.Querier, &limit)
+		if err != nil {
+			return err
+		}
+		if len(servicesVersions) == 0 {
+			results.CheckAfter = serviceCheckInterval
+
+			return nil
+		}
+		if servicesVersions[0].CheckAt.After(time.Now().UTC()) {
+			results.CheckAfter = servicesVersions[0].CheckAt.Sub(time.Now().UTC())
+			if results.CheckAfter < relaxDuration {
+				results.CheckAfter = relaxDuration
+			}
+
+			return nil
+		}
+
+		results.NeedsUpdate = true
+		results.ID = servicesVersions[0].ID
+
+		service, err := models.FindServiceByID(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if service.ServiceType != models.MySQLServiceType {
+			return nil
+		}
+
+		pmmAgents, err := models.FindPMMAgentsForService(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if len(pmmAgents) == 0 {
+			return errors.Errorf("pmmAgent not found for service")
+		}
+		results.AgentID = pmmAgents[0].AgentID
+
+		checkAt := time.Now().UTC().Add(serviceCheckInterval) // TODO
+		if _, err := models.UpdateServiceSoftwareVersions(tx.Querier, servicesVersions[0].ID,
+			models.UpdateServiceSoftwareVersionsParams{CheckAt: &checkAt},
+		); err != nil {
+			return err
+		}
```

Comment:
It seems that we should update this timestamp after the actual versions update, not before.

Reply:
let's have a call regarding this

Reply:
solution: add clarification note